xonsh/proc.py

```diff
@@ -0,0 +1,90 @@
+import io
+import os
+import sys
+
+from threading import Thread
+from subprocess import Popen
+from collections import Sequence
+
+from xonsh.tools import redirect_stdout, redirect_stderr
+
+class ProcProxy(Thread, Popen):
+    def __init__(self, f, args, stdin, stdout, stderr, universal_newlines):
+        self.f = f
+        self.args = args
+        self.pid = None
+        self.returncode = None
+       
+        handles = self._get_handles(stdin, stdout, stderr)
+        (self.p2cread, self.p2cwrite,
+         self.c2pread, self.c2pwrite,
+         self.errread, self.errwrite) = handles
+
+        # default values
+        self.stdin = stdin
+        self.stdout = None
+        self.stderr = None
+
+        if self.p2cwrite != -1:
+            self.stdin = io.open(self.p2cwrite, 'wb', -1)
+            if universal_newlines:
+                self.stdin = io.TextIOWrapper(self.stdin, write_through=True,
+                                              line_buffering=False)
+        if self.c2pread != -1:
+            self.stdout = io.open(self.c2pread, 'rb', -1)
+            if universal_newlines:
+                self.stdout = io.TextIOWrapper(self.stdout)
+        if self.errread != -1:
+            self.stderr = io.open(self.errread, 'rb', -1)
+            if universal_newlines:
+                self.stderr = io.TextIOWrapper(self.stderr)
+
+        Thread.__init__(self)
+        self.start()
+
+    def run(self):
+        if self.f is not None:
```

Comment:
Probably better to say 

```python
if self.f is None:
    return
# need to make file-likes here that work in the "opposite" direction
....
```

That way a level of nesting is saved and the conditional tests a positive.  This is not a blocking issue, but would be nice if you are going back through this part of the code.